File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/request/RequestTest.java
package com.kinvey.androidTest.store.data.request;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.RenamingDelegatingContext;
import android.test.suitebuilder.annotation.SmallTest;

import com.kinvey.android.Client;
import com.kinvey.android.store.DataStore;
import com.kinvey.androidTest.TestManager;
import com.kinvey.androidTest.callback.DefaultKinveyClientCallback;
import com.kinvey.androidTest.callback.DefaultKinveyPushCallback;
import com.kinvey.androidTest.model.Person;
import com.kinvey.java.store.StoreType;

import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;

import static com.kinvey.androidTest.TestManager.PASSWORD;
import static com.kinvey.androidTest.TestManager.TEST_USERNAME;
import static com.kinvey.androidTest.TestManager.USERNAME;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static junit.framework.Assert.assertTrue;

/**
 * Created by yuliya on 1/5/17.
 */

@RunWith(AndroidJUnit4.class)
@SmallTest
public class RequestTest {

    private static final int DEFAULT_TIMEOUT = 60 * 1000;

    private Client client;
    private TestManager<Person> testManager;
    private DataStore<Person> store;

    @Before
    public void setUp() throws InterruptedException, IOException {
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        client = new Client.Builder(mMockContext).build();
        testManager = new TestManager<>();
        testManager.login(USERNAME, PASSWORD, client);

    }

    @After
    public void tearDown() {
        client.performLockDown();
        if (client.getKinveyHandlerThread() != null) {
            try {
                client.stopKinveyHandlerThread();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }


    @Test
    public void testDefaultTimeOutValue() throws InterruptedException, IOException {
        store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        client.enableDebugLogging();
        assertEquals(DEFAULT_TIMEOUT, client.getRequestTimeout());

        Person person = new Person();
        person.setUsername(TEST_USERNAME);
        DefaultKinveyClientCallback callback = testManager.save(store, person);
        assertNotNull(callback);
        assertNotNull(callback.getResult());
        assertNull(callback.getError());
        assertNotNull(callback.getResult().getUsername());
        assertEquals(TEST_USERNAME, callback.getResult().getUsername());

        DefaultKinveyPushCallback pushCallback = testManager.push(store);
        assertNotNull(pushCallback);
        assertNotNull(pushCallback.getResult());
        assertNull(pushCallback.getError());
        assertEquals(1, pushCallback.getResult().getSuccessCount());
    }

    @Test
    @Ignore
    public void testSetCustomTimeoutAndCheckException() throws InterruptedException, IOException {
        store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        assertEquals(DEFAULT_TIMEOUT, client.getRequestTimeout());

        client.setRequestTimeout(1);
        assertEquals(1, client.getRequestTimeout());

        Person person = new Person();
        person.setUsername(TEST_USERNAME);
        DefaultKinveyClientCallback callback = testManager.save(store, person);
        assertNotNull(callback);
        assertNotNull(callback.getResult());
        assertNull(callback.getError());
        assertNotNull(callback.getResult().getUsername());
        assertEquals(TEST_USERNAME, callback.getResult().getUsername());

        DefaultKinveyPushCallback pushCallback = testManager.push(store);
        assertNotNull(pushCallback);
        assertNull(pushCallback.getResult());
        assertNotNull(pushCallback.getError());
        assertEquals("SocketTimeoutException", pushCallback.getError().getClass().getSimpleName());
    }

    @Test
    public void testSetCustomTimeout() throws InterruptedException, IOException {
        store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        assertEquals(DEFAULT_TIMEOUT, client.getRequestTimeout());

        client.setRequestTimeout(DEFAULT_TIMEOUT * 2);
        assertEquals(DEFAULT_TIMEOUT * 2, client.getRequestTimeout());

        Person person = new Person();
        person.setUsername(TEST_USERNAME);
        DefaultKinveyClientCallback callback = testManager.save(store, person);
        assertNotNull(callback);
        assertNotNull(callback.getResult());
        assertNull(callback.getError());
        assertNotNull(callback.getResult().getUsername());
        assertEquals(TEST_USERNAME, callback.getResult().getUsername());

        DefaultKinveyPushCallback pushCallback = testManager.push(store);
        assertNotNull(pushCallback);
        assertNotNull(pushCallback.getResult());
        assertNull(pushCallback.getError());
        assertEquals(1, pushCallback.getResult().getSuccessCount());
    }


}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/data/request/RequestTest.kt
package com.kinvey.androidTest.store.data.request

import android.content.Context
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.model.User
import com.kinvey.android.store.DataStore
import com.kinvey.android.store.DataStore.Companion.collection
import com.kinvey.androidTest.TestManager
import com.kinvey.androidTest.TestManager.Companion.PASSWORD
import com.kinvey.androidTest.TestManager.Companion.TEST_USERNAME
import com.kinvey.androidTest.TestManager.Companion.USERNAME
import com.kinvey.androidTest.model.Person
import com.kinvey.java.store.StoreType
import junit.framework.Assert.*
import org.junit.After
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException

/**
 * Created by yuliya on 1/5/17.
 */

@RunWith(AndroidJUnit4::class)
@SmallTest
class RequestTest {
    private var client: Client<*>? = null
    private var testManager: TestManager<Person>? = null
    private var store: DataStore<Person>? = null
    @Before
    @Throws(InterruptedException::class, IOException::class)
    fun setUp() {
        val mMockContext: Context? = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<User>(mMockContext).build()
        testManager = TestManager()
        testManager?.login(USERNAME, PASSWORD, client)
    }

    @After
    fun tearDown() {
        client?.performLockDown()
        if (Client.kinveyHandlerThread != null) {
            try {
                client?.stopKinveyHandlerThread()
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testDefaultTimeOutValue() {
        store = collection(Person.COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.enableDebugLogging()
        assertEquals(DEFAULT_TIMEOUT, client?.requestTimeout)
        val person = Person()
        person.username = TEST_USERNAME
        val callback = testManager?.save(store, person)
        assertNotNull(callback)
        assertNotNull(callback?.result)
        assertNull(callback?.error)
        assertNotNull(callback?.result?.username)
        assertEquals(TEST_USERNAME, callback?.result?.username)
        val pushCallback = testManager?.push(store)
        assertNotNull(pushCallback)
        assertNotNull(pushCallback?.result)
        assertNull(pushCallback?.error)
        assertEquals(1, pushCallback?.result?.successCount)
    }

    @Test
    @Ignore
    @Throws(InterruptedException::class, IOException::class)
    fun testSetCustomTimeoutAndCheckException() {
        store = collection(Person.COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        assertEquals(DEFAULT_TIMEOUT, client?.requestTimeout)
        client?.requestTimeout = 1
        assertEquals(1, client?.requestTimeout)
        val person = Person()
        person.username = TEST_USERNAME
        val callback = testManager?.save(store, person)
        assertNotNull(callback)
        assertNotNull(callback?.result)
        assertNull(callback?.error)
        assertNotNull(callback?.result?.username)
        assertEquals(TEST_USERNAME, callback?.result?.username)
        val pushCallback = testManager?.push(store)
        assertNotNull(pushCallback)
        assertNull(pushCallback?.result)
        assertNotNull(pushCallback?.error)
        assertEquals("SocketTimeoutException", pushCallback?.error?.javaClass?.simpleName)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSetCustomTimeout() {
        store = collection(Person.COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        assertEquals(DEFAULT_TIMEOUT, client?.requestTimeout)
        client?.requestTimeout = DEFAULT_TIMEOUT * 2
        assertEquals(DEFAULT_TIMEOUT * 2, client?.requestTimeout)
        val person = Person()
        person.username = TEST_USERNAME
        val callback = testManager?.save(store, person)
        assertNotNull(callback)
        assertNotNull(callback?.result)
        assertNull(callback?.error)
        assertNotNull(callback?.result?.username)
        assertEquals(TEST_USERNAME, callback?.result?.username)
        val pushCallback = testManager?.push(store)
        assertNotNull(pushCallback)
        assertNotNull(pushCallback?.result)
        assertNull(pushCallback?.error)
        assertEquals(1, pushCallback?.result?.successCount)
    }

    companion object {
        private const val DEFAULT_TIMEOUT = 60 * 1000
    }
}